dynamic/binary_test.go

```diff
@@ -131,4 +133,45 @@ func TestBinaryUnknownFields(t *testing.T) {
 
 func binaryTranslationParty(t *testing.T, msg proto.Message) {
 	doTranslationParty(t, msg, proto.Marshal, proto.Unmarshal, (*Message).Marshal, (*Message).Unmarshal)
+	binaryTranslationPartyMarshalAppend(t, msg)
+}
+
+func binaryTranslationPartyMarshalAppend(t *testing.T, msg proto.Message) {
+	// Declare a function that has the same interface as (*Message.Marshal) but uses
+	// MarshalAppend internally so we can reuse the translation party tests to verify
+	// the behavior of MarshalAppend in addition to Marshal.
+	marshalAppendSimple := func(m *Message) ([]byte, error) {
+		b := make([]byte, 0, 2048)
+		marshaledB, err := m.MarshalAppend(b)
+
+		// Verify it doesn't allocate a new byte slice.
+		testutil.Eq(t, true, byteSlicesBackedBySameData(b, marshaledB))
+		return marshaledB, err
+	}
+	doTranslationParty(t, msg, proto.Marshal, proto.Unmarshal, marshalAppendSimple, (*Message).Unmarshal)
+
+	marshalAppendPrefix := func(m *Message) ([]byte, error) {
+		var (
+			b      = make([]byte, 0, 2048)
+			prefix = "prefix"
+		)
+		b = append(b, []byte(prefix)...)
```

Comment:
We already check that no new slice is allocated if not necessary above. So we could simplify this case like so:

```go
prefix := "prefix"
marshaledB, err := m.MarshalAppend([]byte(prefix))

// Verify the prefix data is retained.
testutil.Eq(t, prefix, string(marshaledB[:len(prefix)]))
return marshaledB[len(prefix):], err
```